README.md
# profile_photo

Creates a widget that was designed to be used to display a profile photo, but it could also be used in many other ways

## Features

- Ability to add a custom badge to the widget, could be used to display premium users for example
- Will automatically figure out a users initals to display if no image is given, but a name is
- ImageProvider as an image source to use either network or asset images
- onTap and onLongPress call backs
- Adjustable size
- Adjustable corner radius
- Adjustable outline width
- Adjustable background and outline colors

## Getting started

Add the dependency and add anywhere in widget tree

## Usage

Minimum required code for a simple round photo (2nd example)

```dart
ProfilePhoto(
    totalWidth: 100,
    cornerRadius: 80,
    color: Colors.blue,
    image: const AssetImage('assets/test_profile_photo.jpg'),
),
```

More complicated example with initals shown, custom boarder and colors, and badge (3rd example)

```dart
ProfilePhoto(
    totalWidth: 200,
    cornerRadius: 25,
    color: Colors.blue,
    outlineColor: Colors.red,
    outlineWidth: 10,
    textPadding: 10,
    name: 'Brad Varnum',
    fontColor: Colors.white,
    nameDisplayOption: NameDisplayOptions.initials,
    fontWeight: FontWeight.w100,
    showName: true,
    image: const AssetImage('assets/test_profile_photo.jpg'),
    badgeAlignment: Alignment.bottomLeft,
    badgeSize: 60,
    badgeImage: const AssetImage('assets/star.png'),
    onTap: () {
        // open profile for example
    },
    onLongPress: () {
        // popup to message user for example
    },
),
```

<img src="https://raw.githubusercontent.com/bradv123/profile_photo/master/readme_images/example_screenshot.png" alt="example image" width="300"/>

## Additional information

- Feel free to use as-is or fork and modify for what you need
- If you find any issues, please create a issue in this repo
- If you have any suggestions or feature requests, please create an issue
- I made this mostly for fun, and explore the package creation process


CHANGELOG.md
## 1.0.3

* Ran Flutter Format . to pass the static analysis

## 1.0.2

* Updated documentation

## 1.0.1

* Updated readme

## 1.0.0

* Initial Release


test/profile_photo_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:profile_photo/profile_photo.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


lib/src/profile_photo.dart
import 'package:flutter/material.dart';

class ProfilePhoto extends StatelessWidget {
  /// Sets the total width and height of the widget.
  final double totalWidth;

  /// Sets the width of the outer edge of the ProfilePhoto. Default is 0. (inset from totalWidth).
  final int outlineWidth;

  /// Sets the radius of ProfilePhoto. Defualt is 10. (for a circle set equal to totalWidth).
  final double cornerRadius;

  /// Sets the color of the outer edge. If outlineWidth == 0, this doesn't do anything.
  final Color outlineColor;

  /// Sets the main color to show if there is no image.
  final Color color;

  /// onTap callback.
  final VoidCallback? onTap;

  /// onLongPress callback.
  final VoidCallback? onLongPress;

  /// onLongPress callback.
  final VoidCallback? onDoubleTap;

  /// users name, which can be showed or not in various ways using nameDisplayOptions.
  final String name;

  /// Font to use if a name is visible.
  final String? fontFamily;

  /// Font color to use if a name is visible.
  final Color? fontColor;

  /// Sets various ways a name can be shown. Defualt is initials.
  final NameDisplayOptions? nameDisplayOption;

  /// Sets the font weight of the name if visible.
  final FontWeight? fontWeight;

  /// Sets spacing (top, right, bottom & left) around name if its visible. Defualt is 10.
  final int textPadding;

  /// Image to be displayed.
  final ImageProvider? image;

  /// Bool to set if the name should be visible. By default, true if there is no image, false if there is an image set.
  final bool? showName;

  // /// Image to be shown as a badge in the corner of the ProfilePhoto.
  // final ImageProvider? badgeImage;

  /// Widget to be shown as a badge in the corner of the ProfilePhoto.
  final Widget? badge;

  /// Size of the badge. Default is 0.
  // final double badgeSize;

  /// Alignment of the badge. Default is bottom right.
  final Alignment badgeAlignment;

  ProfilePhoto({
    required this.totalWidth,
    required this.color,
    this.outlineWidth = 0,
    this.cornerRadius = 10,
    this.outlineColor = Colors.lightBlue,
    this.onTap,
    this.onLongPress,
    this.onDoubleTap,
    this.name = '',
    this.nameDisplayOption = NameDisplayOptions.initials,
    this.fontColor,
    this.fontFamily,
    this.fontWeight,
    this.textPadding = 10,
    this.image,
    this.showName,
    this.badgeAlignment = Alignment.bottomRight,
    // this.badgeImage,
    this.badge,
    // this.badgeSize = 0,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    String _initials = '';
    String _firstName = '';
    String _lastName = '';
    String _textToShow = '';
    bool _showText = true;
    String _name = name;

    if (_name != '') {
      if (_name.trim().contains(' ')) {
        _name = name.trim();
        _firstName = '${name.substring(0, name.indexOf(' '))}';
        _lastName = '${name.substring(name.indexOf(' ') + 1, name.length)}';
        _initials =
            '${name.substring(0, 1)}${name.substring(name.indexOf(' ') + 1, name.indexOf(' ') + 2)}';
      } else {
        _firstName = name;
        _lastName = name;
        _initials = '${name.substring(0, 1)}';
      }
    } else {
      _initials = ' ';
    }

    switch (nameDisplayOption) {
      case NameDisplayOptions.firstName:
        _textToShow = _firstName;
        break;
      case NameDisplayOptions.lastName:
        _textToShow = _lastName;
        break;
      case NameDisplayOptions.initials:
        _textToShow = _initials;
        break;
      case NameDisplayOptions.dontChange:
        _textToShow = name;
        break;
      case NameDisplayOptions.splitFullName:
        _textToShow = '$_firstName\r\n$_lastName';
        break;
      default:
        _textToShow = ' ';
        break;
    }

    if (image != null) {
      _showText = showName == null ? false : showName!;
    } else if (image == null) {
      _showText = showName == null ? true : showName!;
    }

    return GestureDetector(
      onTap: onTap ?? () {},
      onLongPress: onLongPress ?? () {},
      onDoubleTap: onDoubleTap ?? () {},
      child: Stack(
        alignment: Alignment.center,
        children: [
          // background circle as outline
          Container(
            width: totalWidth,
            height: totalWidth,
            decoration: BoxDecoration(
              borderRadius: BorderRadius.circular(cornerRadius),
              color: outlineColor,
            ),
          ),

          // circle on top for main circle
          Container(
            width: (totalWidth) - (outlineWidth * 2) > 0
                ? (totalWidth) - (outlineWidth * 2)
                : 0,
            height: (totalWidth) - (outlineWidth * 2) > 0
                ? (totalWidth) - (outlineWidth * 2)
                : 0,
            decoration: BoxDecoration(
              borderRadius: cornerRadius - outlineWidth > 0
                  ? BorderRadius.circular(cornerRadius - (outlineWidth))
                  : BorderRadius.circular(cornerRadius),
              color: color,
            ),
            child: image != null
                ? ClipRRect(
                    child: Image(image: image!),
                    borderRadius: cornerRadius - outlineWidth > 0
                        ? BorderRadius.circular(cornerRadius - (outlineWidth))
                        : BorderRadius.circular(cornerRadius),
                  )
                : null,
          ),

          // Top text layer
          if (_showText)
            Container(
              width: (totalWidth) - (outlineWidth * 2) - (textPadding * 2) > 0
                  ? (totalWidth) - (outlineWidth * 2) - (textPadding * 2)
                  : 0,
              height: (totalWidth) - (outlineWidth * 2) - (textPadding * 2) > 0
                  ? (totalWidth) - (outlineWidth * 2) - (textPadding * 2)
                  : 0,
              alignment: Alignment.center,
              child: FittedBox(
                fit: BoxFit.scaleDown,
                child: Text(
                  _textToShow,
                  maxLines: 2,
                  textAlign: TextAlign.center,
                  style: TextStyle(
                    fontSize: 200,
                    fontWeight: fontWeight,
                    color: fontColor,
                    fontFamily: fontFamily,
                  ),
                ),
              ),
            ),
          // if there is a badgeImage to show
          if (badge != null)
            Container(
              width: totalWidth,
              height: totalWidth,
              child: Align(
                alignment: badgeAlignment,
                child: Container(
                  // width: badgeSize,
                  // height: badgeSize,
                  child: badge != null ? badge : null,
                ),
              ),
            ),
        ],
      ),
    );
  }
}

/// Display options for nameDisplayOption
///
/// firstName only shows the first name.
/// lastName only shows the last name.
/// splitFullName shows first name and last name on 2 lines.
/// initals shows the users initals.
/// dontChange shows the name exactly as typed.
enum NameDisplayOptions {
  firstName,
  lastName,
  splitFullName,
  initials,
  dontChange
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:profile_photo/profile_photo.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Profile Photo Example',
      theme: ThemeData(
        primarySwatch: Colors.deepPurple,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('profile_photo example'),
      ),
      body: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Column(
            mainAxisAlignment: MainAxisAlignment.start,
            children: [
              const SizedBox(height: 50),
              ProfilePhoto(
                totalWidth: 75,
                cornerRadius: 75,
                color: Colors.blue,
                outlineColor: Colors.green,
                outlineWidth: 5,
                textPadding: 0,
                name: 'Brad V',
                fontColor: Colors.white,
                nameDisplayOption: NameDisplayOptions.initials,
                fontWeight: FontWeight.w100,
                // badgeSize: 30,
                badgeAlignment: Alignment.bottomRight,
                image: const AssetImage('assets/test_profile_photo.jpg'),
                // badgeImage: const AssetImage('assets/star.png'),
                badge: Container(
                  height: 25,
                  width: 25,
                  decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(200),
                    color: Colors.grey,
                  ),
                  alignment: Alignment.center,
                  child: Icon(
                    Icons.add_alarm_outlined,
                    size: 15,
                    color: Colors.white,
                  ),
                ),
                onTap: () {
                  // open profile
                },
                onLongPress: () {
                  // popup to message user
                },
              ),
              const SizedBox(height: 50),
              ProfilePhoto(
                totalWidth: 100,
                cornerRadius: 80,
                color: Colors.blue,
                image: const AssetImage('assets/test_profile_photo.jpg'),
              ),
              const SizedBox(height: 50),
              ProfilePhoto(
                totalWidth: 200,
                cornerRadius: 25,
                color: Colors.blue,
                outlineColor: Colors.red,
                outlineWidth: 10,
                textPadding: 10,
                name: 'Brad Varnum',
                fontColor: Colors.white,
                nameDisplayOption: NameDisplayOptions.initials,
                fontWeight: FontWeight.w100,
                showName: true,
                image: const AssetImage('assets/test_profile_photo.jpg'),
                badgeAlignment: Alignment.bottomLeft,
                // badgeSize: 60,
                // badgeImage: const AssetImage('assets/star.png'),
                badge: const Text('Badge'),
                onTap: () {
                  // open profile
                },
                onLongPress: () {
                  // popup to message user
                },
              ),
              const SizedBox(height: 50),
              ProfilePhoto(
                totalWidth: 150,
                cornerRadius: 20,
                color: Colors.deepPurple.shade300,
                outlineColor: Colors.cyan,
                outlineWidth: 10,
                textPadding: 15,
                name: 'Brad Varnum',
                fontColor: Colors.white,
                nameDisplayOption: NameDisplayOptions.splitFullName,
                fontWeight: FontWeight.w500,
                showName: true,
              ),
              const SizedBox(height: 0),
            ],
          ),
        ],
      ),
    );
  }
}
